suricata.yaml.in

```diff
@@ -163,6 +163,9 @@ outputs:
       #  pipelining:
       #    enabled: yes ## set enable to yes to enable query pipelining
       #    batch-size: 10 ## number of entry to keep in buffer
+      metadata:
```

Comment:
@jasonish does this fix your ideas for a eve metadata config option?

Reply:
I was thinking it could be a boolean value or a list of metadata to log:
```
metadata: yes/no

metadata: [flowbits, flowints]
```
where
```
metadata:
```
would be the equivalent of an empty list, so off.

It doesn't look like you are putting this filename under the metadata object though. So should it just be logged if available?.

Reply:
I suppose empty could default to all as well.

Reply:
Don't think the pcap name should be logged unconditionally, or by default. Paths can be long so it would add a lot of unnecessary storage for those that don't need it.

Reply:
This gets tricky then. I think stuff like flowbits should be logged by default. Which means "all" metadata on by default is not desirable if we do want stuff like flowbits by default, but not the pcap_filename.

Reply:
metadata: [defaults] ? Would pick a reasonable default.
metadata: [all] everything
metadata: [flowbits, pcap-filename] just flowbits and pcap-filename.
(forget the syntax in this example)

would this make sense?

Reply:
I think this looks better:
```
metadata: defaults
metadata: all
metadata: no # or off, false
```
Where a scalar value has special meaning. And a list is a literal list of items:
```
metadata: [flowbits, pcap-filename]
```
However, there might be a use case for default plus something. Then maybe a special naming convention is needed?
```
metadata:
  - _defaults_
  - pcap-filename
```

This does mean that metadata can be a scalar or a list. But thats pretty easy to deal with an test at runtime.

Reply:
does this need to be a list at all? A map sounds simpler to me.

Reply:
I map wouldn't look very natural:
```
metadata:
  pcap-filename:
  flowbits:
```
where the list woould be:
```
metadata: [pcap-filename, flowbits]
```
or
```
metadata:
  - pcap
  - filename
```

Reply:
See edit above.

Reply:
OK. My thinking was mostly to use lists where it makes sense to have things multiple times, and maps where it doesn't.
But I'm fine with making it a list if you think it makes sense. Can't get over the magic ``_keyword_`` notation though.

Reply:
I think its better than simply "all" and "defaults" as magic, as this way its clear they have special meaning. But the words all and defaults are generic enough that they could become magic reserved words in this context.

Reply:
I think the underscore variants look way more unnatural that the map would be, but maybe thats just me :)

Reply:
> I think the underscore variants look way more unnatural that the map would be, but maybe thats just me :)

Not sure what you mean. I used underscore variants in the map?

Reply:
So lets look at what we want to express.  We want no metadata, all metadata, or a special selection we choose as the default set. Or selected set of meta fields?

I think the most straightforward way is:
```
metadata: false
metadata: defaults
metadata: all
```
This values all have a specific meaning. If no metadata field is provided, or is empty, default to "defaults".

But if I want specific values, then I have to choose:
```
metadata:
- flowbits
- pcap-filename
```
No values with magic meaning. All possible values map to a real field in the output. Think of them as a set, and in the code treat it like a set. As they are not maps we don't have to think about what their value is. 

Or a map:
```
metadata:
  flowbits: true
  pcap-filename: true
```
An empty value is generally considered falsey, so if a map I think we should think about the values.

Reply:
Do we have other examples of fields that can be both having a single value or a list (or map)? What would it mean for overriding from the commandline? E.g. if the yaml has a list, and I --set it to 'all'?

Reply:
Not that I can see. But its pretty easy to do. Code could be something like:
```
        /* Check if top-level metadata should be logged. */
        const ConfNode *metadata = ConfNodeLookupChild(conf, "metadata");
        if (metadata != NULL) {
            if (metadata->val == NULL && metadata->is_seq) {
                SCLogNotice("We have a metadata list.");
            } else if (metadata->val) {
                SCLogNotice("We have a mdatadata value.");
            }
        }
```
So even we have:
```
metadata: [a, b, c]
```
Doing a "--set outputs.1.eve-log.metadata=all" will still take precedence as we check for a value before considering it a list.